dragonflybsd/DragonFlyBSDProcessList.c

```diff
@@ -261,29 +261,68 @@ static inline void DragonFlyBSDProcessList_scanMemoryInfo(ProcessList* pl) {
    pl->usedSwap *= pageSizeKb;
 }
 
-static char* DragonFlyBSDProcessList_readProcessName(kvm_t* kd, const struct kinfo_proc* kproc, int* basenameEnd) {
+//static void DragonFlyBSDProcessList_updateExe(const struct kinfo_proc* kproc, Process* proc) {
+//   const int mib[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, kproc->kp_pid };
+//   char buffer[2048];
+//   size_t size = sizeof(buffer);
+//   if (sysctl(mib, 4, buffer, &size, NULL, 0) != 0) {
+//      Process_updateExe(proc, NULL);
+//      return;
+//   }
+//
+//   /* Kernel threads return an empty buffer */
+//   if (buffer[0] == '\0') {
+//      Process_updateExe(proc, NULL);
+//      return;
+//   }
+//
+//   Process_updateExe(proc, buffer);
+//}
+
+static void DragonFlyBSDProcessList_updateExe(const struct kinfo_proc* kproc, Process* proc) {
+   if (Process_isKernelThread(proc))
+      return;
+
+   char path[PATH_MAX];
+   xSnprintf(path, sizeof(path), "/proc/%d/file", kproc->kp_pid);
+
+   char target[PATH_MAX];
+   ssize_t ret = readlink(path, target, sizeof(target) - 1);
+   if (ret <= 0)
+      return;
+
+   target[ret] = '\0';
+   Process_updateExe(proc, target);
+}
+
+static void DragonFlyBSDProcessList_updateProcessName(kvm_t* kd, const struct kinfo_proc* kproc, Process* proc) {
+   Process_updateComm(proc, kproc->kp_comm);
+
    char** argv = kvm_getargv(kd, kproc, 0);
-   if (!argv) {
-      return xStrdup(kproc->kp_comm);
+   if (!argv || !argv[0]) {
+      Process_updateCmdline(proc, kproc->kp_comm, 0, strlen(kproc->kp_comm));
+      return;
    }
-   int len = 0;
+
+   size_t len = 0;
    for (int i = 0; argv[i]; i++) {
       len += strlen(argv[i]) + 1;
    }
-   char* comm = xMalloc(len);
-   char* at = comm;
-   *basenameEnd = 0;
+
+   char* cmdline = xMalloc(len);
+   char* at = cmdline;
+   int end = 0;
    for (int i = 0; argv[i]; i++) {
       at = stpcpy(at, argv[i]);
-      if (!*basenameEnd) {
-         *basenameEnd = at - comm;
+      if (end == 0) {
+         end = at - cmdline;
       }
-      *at = ' ';
-      at++;
+      *at++ = ' ';
    }
    at--;
    *at = '\0';
-   return comm;
+
+   Process_updateCmdline(proc, cmdline, 0, end);
```

Comment:
Potentially highlights full program path instead of just program basename inside cmdline.

Reply:
added via `skipPotentialPath()` into `Process_updateCmdline()`